File: bigmali/prior.py
import numpy as np
import scipy.interpolate as interpolate
import hmf

# class DataPrior(object):
#     def __init__(self, mass):
#         space = np.logspace(mass.min(), mass.max(), 1000)
#         x,y = np.histogram(mass, space, normed=True)
#         self.mass = np.append(np.append(, mass))
#         probex = np.append(np.append(, y))
#         self.prob = probex / np.trapz(probex, x=self.mass)
#         cumsum = np.cumsum(self.prob) / np.sum(self.prob)
#         self.inv_cdf = interpolate.interp1d(cumsum, self.mass)
#
#     def logpdf(self, mass):
#         return np.log(self.pdf(mass))
#
#     def pdf(self, mass):
#         """
#         Note: Assumes the mass is within range of the prior.
#         """
#         right_ind = np.minimum(np.searchsorted(self.mass, mass), len(self.mass) - 1)
#         left_ind = right_ind - 1
#         # find where we fall in interval between masses (similar to trapezoidal integration)
#         f = (mass - self.mass[left_ind]) / (self.mass[right_ind] - self.mass[left_ind])
#         return f * self.prob[right_ind] + (1 - f) * self.prob[left_ind]
#
#     def rvs(self, size=1):
#         return self.inv_cdf(np.random.rand(size))

class MassPrior(object):
    """
    Class to represent mass prior distribution.
    Initialize with mass and probability arrays.
    Can both evaluate and sample from distribution.
    """
    def __init__(self, mass, prob):
        self.prob = prob
        # add epsilon background
        epsilon = 1e-30
        self.mass = np.append(np.append([1, mass.min()-1], mass), [mass.max()+1, 100 * mass.max()])
        deps = [epsilon, epsilon]
        probex = np.append(np.append(deps, prob), deps)
        self.prob = probex / np.trapz(probex, x=self.mass)
        # need to normalize by prob values for interpolation because normalized against
        # trapezoidal integration
        cumsum = np.cumsum(self.prob) / np.sum(self.prob)
        self.inv_cdf = interpolate.interp1d(cumsum, self.mass)

    def logpdf(self, mass):
        return np.log(self.pdf(mass))

    def pdf(self, mass):
        """
        Note: Assumes the mass is within range of the prior.
        """
        right_ind = np.minimum(np.searchsorted(self.mass, mass), len(self.mass) - 1)
        left_ind = right_ind - 1
        # find where we fall in interval between masses (similar to trapezoidal integration)
        f = (mass - self.mass[left_ind]) / (self.mass[right_ind] - self.mass[left_ind])
        return f * self.prob[right_ind] + (1 - f) * self.prob[left_ind]

    def rvs(self, size=1):
        return self.inv_cdf(np.random.rand(size))

    @staticmethod
    def default(h=0.73, mmin=10.2358590918, mmax=14.3277327776, z=0):
        mf = hmf.MassFunction(z=z, Mmin=mmin, Mmax=mmax,
                              cosmo_model=hmf.cosmo.WMAP5,
                              hmf_model=hmf.fitting_functions.Tinker10)
        return MassPrior(mf.m * h, mf.dndm)


class TinkerPrior(object):
    """
    Class that assigns tinker prior (Tinker10 from hmf) to each redshift bin in grid.
    """
    def __init__(self, grid, h=0.73, mmin=10.2358590918, mmax=14.3277327776):
        self.grid = grid
        self.grid_to_prior = dict()
        self.min_mass = float("inf")
        self.max_mass = 0
        for z in grid.redshifts:
            mf = hmf.MassFunction(z=z, Mmin=mmin, Mmax=mmax,
                                  cosmo_model=hmf.cosmo.WMAP5,
                                  hmf_model=hmf.fitting_functions.Tinker10)
            self.grid_to_prior[z] = MassPrior(mf.m * h, mf.dndm / np.trapz(mf.dndm, x=mf.m * h))
            self.min_mass = min(np.min((mf.m * h).min()), self.min_mass)
            self.max_mass = max((mf.m * h).max(), self.max_mass)

    def fetch(self, z):
        return self.grid_to_prior[self.grid.snap(z)]

    def pdf(self, mass, z):
        return self.fetch(z).pdf(mass)

    def logpdf(self, mass, z):
        return self.fetch(z).logpdf(mass)

    def rvs(self, z, size=1):
        return self.fetch(z).rvs(size=size)

File: src/prior.cpp
#include "prior.hpp"
#include <algorithm>
#include <iostream>
#include <cmath>
#include <fstream>

double MassPrior::trapz(const std::vector<double>& x, const std::vector<double>& y)
{
	double sum, avg_height, width;
	sum = 0;
	for (int i=0; i < x.size()-1; i++)
	{
		avg_height = (0.5 * (y[i] + y[i+1]));
		width = x[i+1] - x[i];
		sum = sum + width * avg_height;
	}
	return sum;
}

MassPrior::MassPrior(std::vector<double>& masses, std::vector<double>& probs)
{
	const double epsilon = 1e-30;
	std::vector<double> mass(410);
	std::vector<double> prob(410);
	mass = masses;
	prob = probs;

	mass.insert(mass.begin(), mass[0] - 1);
	mass.insert(mass.begin(), 1);
	mass.push_back(mass.back() + 1);
	mass.push_back(mass.back() * 100);

	prob.insert(prob.begin(), epsilon);
	prob.insert(prob.begin(), epsilon);
	prob.push_back(epsilon);
	prob.push_back(epsilon);


	const int n = prob.size();
	const double norm_trapz = trapz(mass, prob);
	double norm_sum = 0;
	std::vector<double> cumsum = std::vector<double>(n);
	for (int i=0; i<n; i++)
	{
		prob[i] = prob[i] / norm_trapz;
		norm_sum += prob[i];
	}
	cumsum[0] = 0;
	for (int i=1; i<n; i++)
	{
		cumsum[i] = cumsum[i-1] + (prob[i] / norm_sum);
	}

	this->mass = mass;
	this->prob = prob;
	this->cumsum = cumsum;

    std::uniform_real_distribution<double> dist(0, 1);
    this->dist = dist;
}

double MassPrior::pdf(const double mass)
{
	auto upper = std::upper_bound(this->mass.begin(), this->mass.end(), mass);
	int right_ind = upper - this->mass.begin();
	int left_ind = right_ind - 1;
	double f = (mass - this->mass[left_ind]) / (this->mass[right_ind] - this->mass[left_ind]);
	return f * this->prob[right_ind] + (1-f) * this->prob[left_ind];
}

double MassPrior::logpdf(const double mass)
{
	return this->pdf(mass);
}

void MassPrior::rvs(std::vector<double>& samples)
{
	int n = samples.size();
	for (int i=0;i<n;i++)
	{
		double r = this->dist(this->gen);
		auto upper = std::upper_bound(this->cumsum.begin(), this->cumsum.end(), r);
		int right_ind = upper - this->cumsum.begin();
		int left_ind = right_ind - 1;
		double f = (r - this->cumsum[left_ind]) / (this->cumsum[right_ind] - this->cumsum[left_ind]);
		samples[i] = f * this->mass[right_ind] + (1-f) * this->mass[left_ind];
	}
}

void MassPrior::print()
{
	int n = this->mass.size();
	for (int i=0; i<n; i++)
	{
		std::cout << this->mass[i] << ","\
		<< this->prob[i] << ","\
		<< this->cumsum[i] << std::endl;
	}
}

TinkerPrior::TinkerPrior()
{
	double z, m, p;
	std::ifstream infile("massprior.txt");
	this->priormap = std::map<long, MassPrior>();
	std::vector<double> mass(410);
	std::vector<double> prob(410);

	infile >> z >> m >> p;
	double prevz = z;
	int i = 0;
	mass[i] = m;
	prob[i] = p;
	i += 1;
	while (infile >> z >> m >> p)
	{
		if (z != prevz)
		{
			this->priormap.emplace(TinkerPrior::snap(prevz), MassPrior(mass, prob));
			i = 0;
			mass[i] = m;
			prob[i] = p;
			i += 1;
			prevz = z;
		}
		else{
			mass[i] = m;
			prob[i] = p;
			i += 1;
			// std::cout >> z >> "," >> mass >> "," >> prob >> std::endl;			
		}
	}
	this->priormap.emplace(TinkerPrior::snap(z), MassPrior(mass, prob));
}

double TinkerPrior::pdf(double z, const double mass)
{
	return this->priormap.find(TinkerPrior::snap(z))->second.pdf(mass);
}

double TinkerPrior::logpdf(double z, const double mass)
{
	return this->priormap.find(TinkerPrior::snap(z))->second.logpdf(mass);
}

void TinkerPrior::rvs(double z, std::vector<double>& samples)
{
	auto search = this->priormap.find(TinkerPrior::snap(z));
	search->second.rvs(samples);
}

MassPrior TinkerPrior::fetch(double z)
{
	return this->priormap.find(TinkerPrior::snap(z))->second;
}

long TinkerPrior::snap(const double z)
{
	return (long) z * 100;
}
